### PR TITLE
auth-server: server: rate-limiter: Implement bundle rate limits

### PR DESCRIPTION
**Pull request by joeykraut**
### Purpose
This PR implements bundle rate limits for the auth server. Specifically, we rate limit the number of inflight, unsettled bundles per client. So, a client may accrue new API tokens by either:
- Waiting for their token bucket to refill
- Settling a match

### Testing
- [ ] Test in testnet

### CODE REVIEW

File: auth/auth-server/src/server/rate_limiter.rs [hunk 1]
```diff
@@ -0,0 +1,70 @@
+//! A rate limiter for the server
+//!
+//! The unit which we rate limit is number of inflight bundles. Therefore, there
+//! are two ways for the token bucket to refill:
+//!     - Wait for the next refill
+//!     - Settle a bundle on-chain
+//!
+//! The ladder is measured by waiting for nullifier spend events on-chain
```
Comment (sehyunc) [inline auth/auth-server/src/server/rate_limiter.rs:8]: *latter

**Review by sehyunc — APPROVED** [review]
clean

File: auth/auth-server/src/server/handle_external_match.rs [hunk 2]
```diff
@@ -67,36 +79,79 @@ impl Server {
...
+
+        // If the bundle settles, increase the API user's a rate limit token balance
+        let did_settle = await_settlement(&match_resp.match_bundle, &self.arbitrum_client).await?;
+        if did_settle {
+            self.add_rate_limit_token(key.clone()).await;
+        }
+
+        // Log the bundle and record metrics
+        self.log_bundle(resp)?;
+        record_external_match_metrics(&order, match_resp, key, did_settle).await
```
Comment (sehyunc) [inline auth/auth-server/src/server/handle_external_match.rs:152]: I'm assuming we're ok with using the same metric name for external match and quote and assemble endpoints?

Comment (joeykraut) [reply auth/auth-server/src/server/handle_external_match.rs:152]: yeah i think we care more about the aggregate

**Pull request merged**